File: services/review-service/src/modules/reviews/services/review.service.ts
import {
  BadRequestException,
  ConflictException,
  ForbiddenException,
  Injectable,
  NotFoundException,
  UnprocessableEntityException
} from '@nestjs/common';
import { ErrorCode } from '../../../common/constants/error-code.enum';
import { Role } from '../../../common/constants/role.enum';
import { AuthenticatedUserContext } from '../../../common/types/request-context.type';
import { CreateReviewDto, ListReviewsDto, ModerateReviewDto, ReplyReviewDto, UpdateReviewDto } from '../dto';
import { ReviewDocument } from '../entities/review.entity';
import { ReviewStatus } from '../enums/review-status.enum';
import { ReviewRepository } from '../repositories/review.repository';

const MODERATOR_ROLES: Role[] = [Role.ADMIN, Role.SUPPORT, Role.SUPER_ADMIN];
const REPLY_ROLES: Role[] = [Role.SELLER, Role.ADMIN, Role.SUPPORT, Role.SUPER_ADMIN];

@Injectable()
export class ReviewService {
  constructor(private readonly reviewRepository: ReviewRepository) {}

  async createReview(user: AuthenticatedUserContext, dto: CreateReviewDto): Promise<Record<string, unknown>> {
    this.assertRole(user, [Role.CUSTOMER]);

    const existing = await this.reviewRepository.findActiveDuplicate(dto.orderId, dto.productId, user.userId);
    if (existing) {
      throw new ConflictException({
        code: ErrorCode.REVIEW_ALREADY_EXISTS,
        message: 'Review already exists for this order and product'
      });
    }

    const review = await this.reviewRepository.create({
      orderId: dto.orderId,
      productId: dto.productId,
      sellerId: dto.sellerId,
      buyerId: user.userId,
      rating: dto.rating,
      title: dto.title?.trim() || undefined,
      content: dto.content.trim(),
      images: dto.images ?? [],
      status: ReviewStatus.PUBLISHED
    });

    return this.toReviewResponse(review);
  }

  async listReviews(user: AuthenticatedUserContext | undefined, query: ListReviewsDto): Promise<Record<string, unknown>> {
    this.assertBuyerFilterPermission(user, query);

    const status = this.resolveStatusFilter(user, query);
    const { items, totalItems } = await this.reviewRepository.list({
      status,
      query
    });

    const page = query.page ?? 1;
    const pageSize = query.pageSize ?? 20;

    return {
      items: items.map((item) => this.toReviewResponse(item)),
      pagination: {
        page,
        pageSize,
        totalItems,
        totalPages: Math.ceil(totalItems / pageSize)
      }
    };
  }

  async getReviewById(user: AuthenticatedUserContext | undefined, reviewId: string): Promise<Record<string, unknown>> {
    const review = await this.reviewRepository.findById(reviewId);
    this.assertReviewExists(review);

    if (!this.canViewReview(user, review!)) {
      throw new NotFoundException({
        code: ErrorCode.REVIEW_NOT_FOUND,
        message: 'Review not found'
      });
    }

    return this.toReviewResponse(review!);
  }

  async updateReview(user: AuthenticatedUserContext, reviewId: string, dto: UpdateReviewDto): Promise<Record<string, unknown>> {
    this.assertRole(user, [Role.CUSTOMER]);

    const review = await this.reviewRepository.findById(reviewId);
    this.assertReviewExists(review);

    if (review!.buyerId !== user.userId) {
      throw new ForbiddenException({
        code: ErrorCode.FORBIDDEN,
        message: 'You can only update your own review'
      });
    }

    const updated = await this.reviewRepository.updateById(reviewId, {
      rating: dto.rating ?? review!.rating,
      title: dto.title?.trim() ?? review!.title,
      content: dto.content?.trim() ?? review!.content,
      images: dto.images ?? review!.images
    });

    this.assertReviewExists(updated);
    return this.toReviewResponse(updated!);
  }

  async deleteReview(user: AuthenticatedUserContext, reviewId: string): Promise<Record<string, unknown>> {
    this.assertRole(user, [Role.CUSTOMER]);

    const review = await this.reviewRepository.findById(reviewId);
    this.assertReviewExists(review);

    if (review!.buyerId !== user.userId) {
      throw new ForbiddenException({
        code: ErrorCode.FORBIDDEN,
        message: 'You can only delete your own review'
      });
    }

    const updated = await this.reviewRepository.updateById(reviewId, {
      status: ReviewStatus.DELETED,
      deletedAt: new Date()
    });

    if (!updated) {
      throw new NotFoundException({
        code: ErrorCode.REVIEW_NOT_FOUND,
        message: 'Review not found'
      });
    }

    return this.toReviewResponse(updated!);
  }

  async moderateReview(user: AuthenticatedUserContext, reviewId: string, dto: ModerateReviewDto): Promise<Record<string, unknown>> {
    this.assertRole(user, MODERATOR_ROLES);

    if (dto.status === ReviewStatus.DELETED) {
      throw new BadRequestException({
        code: ErrorCode.BAD_REQUEST,
        message: 'Moderation status cannot be DELETED'
      });
    }

    if ([ReviewStatus.HIDDEN, ReviewStatus.REJECTED].includes(dto.status) && !dto.reason?.trim()) {
      throw new UnprocessableEntityException({
        code: ErrorCode.REVIEW_MODERATION_REASON_REQUIRED,
        message: 'Moderation reason is required for HIDDEN or REJECTED status'
      });
    }

    const review = await this.reviewRepository.findById(reviewId);
    this.assertReviewExists(review);

    const updated = await this.reviewRepository.updateById(reviewId, {
      status: dto.status,
      moderationReason: dto.reason?.trim(),
      moderatedBy: user.userId,
      moderatedAt: new Date()
    });

    this.assertReviewExists(updated);
    return this.toReviewResponse(updated!);
  }

  async replyReview(user: AuthenticatedUserContext, reviewId: string, dto: ReplyReviewDto): Promise<Record<string, unknown>> {
    this.assertRole(user, REPLY_ROLES);

    const review = await this.reviewRepository.findById(reviewId);
    this.assertReviewExists(review);

    if (review!.status === ReviewStatus.DELETED) {
      throw new NotFoundException({
        code: ErrorCode.REVIEW_NOT_FOUND,
        message: 'Review not found'
      });
    }

    if (user.role === Role.SELLER && review!.sellerId !== user.userId) {
      throw new ForbiddenException({
        code: ErrorCode.FORBIDDEN,
        message: 'Seller can only reply to own product reviews'
      });
    }

    const updated = await this.reviewRepository.updateById(reviewId, {
      reply: {
        content: dto.content.trim(),
        repliedBy: user.userId,
        repliedAt: new Date()
      }
    });

    this.assertReviewExists(updated);
    return this.toReviewResponse(updated!);
  }

  async getProductSummary(productId: string): Promise<Record<string, unknown>> {
    const summary = await this.reviewRepository.getProductSummary(productId);

    return {
      productId,
      averageRating: summary.averageRating,
      totalReviews: summary.totalReviews,
      starDistribution: summary.starDistribution
    };
  }

  private resolveStatusFilter(user: AuthenticatedUserContext | undefined, query: ListReviewsDto): ReviewStatus | undefined {
    if (!user) {
      return ReviewStatus.PUBLISHED;
    }

    if (MODERATOR_ROLES.includes(user.role)) {
      return query.status;
    }

    if (user.role === Role.CUSTOMER && query.buyerId === user.userId) {
      return query.status;
    }

    if (query.status && query.status !== ReviewStatus.PUBLISHED) {
      throw new ForbiddenException({
        code: ErrorCode.FORBIDDEN,
        message: 'Insufficient role to query this review status'
      });
    }

    return ReviewStatus.PUBLISHED;
  }

  private assertBuyerFilterPermission(user: AuthenticatedUserContext | undefined, query: ListReviewsDto): void {
    if (!query.buyerId) {
      return;
    }

    if (!user) {
      throw new ForbiddenException({
        code: ErrorCode.FORBIDDEN,
        message: 'Authentication required for buyer filter'
      });
    }

    if (MODERATOR_ROLES.includes(user.role)) {
      return;
    }

    if (user.role === Role.CUSTOMER && query.buyerId === user.userId) {
      return;
    }

    throw new ForbiddenException({
      code: ErrorCode.FORBIDDEN,
      message: 'Insufficient role to filter by buyerId'
    });
  }

  private canViewReview(user: AuthenticatedUserContext | undefined, review: ReviewDocument): boolean {
    if (review.status === ReviewStatus.PUBLISHED) {
      return true;
    }

    if (review.status === ReviewStatus.DELETED) {
      return false;
    }

    if (!user) {
      return false;
    }

    if (MODERATOR_ROLES.includes(user.role)) {
      return true;
    }

    if (user.role === Role.CUSTOMER && review.buyerId === user.userId) {
      return true;
    }

    if (user.role === Role.SELLER && review.sellerId === user.userId) {
      return true;
    }

    return false;
  }

  private assertRole(user: AuthenticatedUserContext, expectedRoles: Role[]): void {
    if (!expectedRoles.includes(user.role)) {
      throw new ForbiddenException({
        code: ErrorCode.FORBIDDEN,
        message: 'Insufficient role'
      });
    }
  }

  private assertReviewExists(review: ReviewDocument | null): void {
    if (!review || review.status === ReviewStatus.DELETED) {
      throw new NotFoundException({
        code: ErrorCode.REVIEW_NOT_FOUND,
        message: 'Review not found'
      });
    }
  }

  private toReviewResponse(review: ReviewDocument): Record<string, unknown> {
    return {
      id: review.id,
      orderId: review.orderId,
      productId: review.productId,
      sellerId: review.sellerId,
      buyerId: review.buyerId,
      rating: review.rating,
      title: review.title ?? null,
      content: review.content,
      images: review.images ?? [],
      status: review.status,
      moderationReason: review.moderationReason ?? null,
      moderatedBy: review.moderatedBy ?? null,
      moderatedAt: review.moderatedAt?.toISOString() ?? null,
      reply: review.reply
        ? {
            content: review.reply.content,
            repliedBy: review.reply.repliedBy,
            repliedAt: review.reply.repliedAt.toISOString()
          }
        : null,
      deletedAt: review.deletedAt?.toISOString() ?? null,
      createdAt: review.createdAt.toISOString(),
      updatedAt: review.updatedAt.toISOString()
    };
  }
}

File: services/review-service/internal/service/review_service.go
package service

import (
	"context"
	"net/http"
	"strings"
	"time"

	"review-service-go/internal/domain"
	"review-service-go/internal/httpx"
	"review-service-go/internal/repository"

	"go.mongodb.org/mongo-driver/mongo"
)

var moderatorRoles = map[domain.Role]struct{}{
	domain.RoleAdmin:      {},
	domain.RoleSupport:    {},
	domain.RoleSuperAdmin: {},
}

var replyRoles = map[domain.Role]struct{}{
	domain.RoleSeller:     {},
	domain.RoleAdmin:      {},
	domain.RoleSupport:    {},
	domain.RoleSuperAdmin: {},
}

type ReviewService interface {
	CreateReview(ctx context.Context, user domain.UserContext, input domain.CreateReviewInput) (ReviewResponse, error)
	ListReviews(ctx context.Context, user *domain.UserContext, query domain.ListReviewsQuery) (ListReviewsResponse, error)
	GetReviewByID(ctx context.Context, user *domain.UserContext, reviewID string) (ReviewResponse, error)
	UpdateReview(ctx context.Context, user domain.UserContext, reviewID string, input domain.UpdateReviewInput) (ReviewResponse, error)
	DeleteReview(ctx context.Context, user domain.UserContext, reviewID string) (ReviewResponse, error)
	ModerateReview(ctx context.Context, user domain.UserContext, reviewID string, input domain.ModerateReviewInput) (ReviewResponse, error)
	ReplyReview(ctx context.Context, user domain.UserContext, reviewID string, input domain.ReplyReviewInput) (ReviewResponse, error)
	GetProductSummary(ctx context.Context, productID string) (domain.ProductSummary, error)
}

type reviewService struct {
	repo repository.ReviewRepository
}

func NewReviewService(repo repository.ReviewRepository) ReviewService {
	return &reviewService{repo: repo}
}

func (s *reviewService) CreateReview(ctx context.Context, user domain.UserContext, input domain.CreateReviewInput) (ReviewResponse, error) {
	if user.Role != domain.RoleCustomer {
		return ReviewResponse{}, httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "Insufficient role", nil)
	}

	existing, err := s.repo.FindActiveDuplicate(ctx, input.OrderID, input.ProductID, user.UserID)
	if err != nil {
		return ReviewResponse{}, err
	}
	if existing != nil {
		return ReviewResponse{}, httpx.NewAppError(http.StatusConflict, domain.ErrorCodeReviewAlreadyExists, "Review already exists for this order and product", nil)
	}

	title := normalizeCreateTitle(input.Title)
	content := strings.TrimSpace(input.Content)

	review, err := s.repo.Create(ctx, repository.CreateReviewPayload{
		OrderID:   input.OrderID,
		ProductID: input.ProductID,
		SellerID:  input.SellerID,
		BuyerID:   user.UserID,
		Rating:    input.Rating,
		Title:     title,
		Content:   content,
		Images:    nonNilStringSlice(input.Images),
		Status:    domain.ReviewStatusPublished,
	})
	if err != nil {
		if mongo.IsDuplicateKeyError(err) {
			return ReviewResponse{}, httpx.NewAppError(http.StatusConflict, domain.ErrorCodeReviewAlreadyExists, "Review already exists for this order and product", nil)
		}
		return ReviewResponse{}, err
	}

	return toReviewResponse(review), nil
}

func (s *reviewService) ListReviews(ctx context.Context, user *domain.UserContext, query domain.ListReviewsQuery) (ListReviewsResponse, error) {
	if err := assertBuyerFilterPermission(user, query); err != nil {
		return ListReviewsResponse{}, err
	}

	status, err := resolveStatusFilter(user, query)
	if err != nil {
		return ListReviewsResponse{}, err
	}
	query.Status = status

	items, totalItems, err := s.repo.List(ctx, query)
	if err != nil {
		return ListReviewsResponse{}, err
	}

	responses := make([]ReviewResponse, 0, len(items))
	for _, item := range items {
		responses = append(responses, toReviewResponse(item))
	}

	totalPages := int64(1)
	if query.PageSize > 0 {
		totalPages = (totalItems + int64(query.PageSize) - 1) / int64(query.PageSize)
		if totalPages < 1 {
			totalPages = 1
		}
	}

	return ListReviewsResponse{
		Items: responses,
		Pagination: Pagination{
			Page:       query.Page,
			PageSize:   query.PageSize,
			TotalItems: totalItems,
			TotalPages: totalPages,
		},
	}, nil
}

func (s *reviewService) GetReviewByID(ctx context.Context, user *domain.UserContext, reviewID string) (ReviewResponse, error) {
	review, err := s.repo.FindByID(ctx, reviewID)
	if err != nil {
		return ReviewResponse{}, err
	}
	if review == nil || review.Status == domain.ReviewStatusDeleted {
		return ReviewResponse{}, httpx.NewAppError(http.StatusNotFound, domain.ErrorCodeReviewNotFound, "Review not found", nil)
	}

	if !canViewReview(user, *review) {
		return ReviewResponse{}, httpx.NewAppError(http.StatusNotFound, domain.ErrorCodeReviewNotFound, "Review not found", nil)
	}

	return toReviewResponse(*review), nil
}

func (s *reviewService) UpdateReview(ctx context.Context, user domain.UserContext, reviewID string, input domain.UpdateReviewInput) (ReviewResponse, error) {
	if user.Role != domain.RoleCustomer {
		return ReviewResponse{}, httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "Insufficient role", nil)
	}

	review, err := s.repo.FindByID(ctx, reviewID)
	if err != nil {
		return ReviewResponse{}, err
	}
	if review == nil || review.Status == domain.ReviewStatusDeleted {
		return ReviewResponse{}, httpx.NewAppError(http.StatusNotFound, domain.ErrorCodeReviewNotFound, "Review not found", nil)
	}
	if review.BuyerID != user.UserID {
		return ReviewResponse{}, httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "You can only update your own review", nil)
	}

	rating := review.Rating
	if input.Rating != nil {
		rating = *input.Rating
	}
	title := review.Title
	if input.Title != nil {
		t := strings.TrimSpace(*input.Title)
		title = &t
	}
	content := review.Content
	if input.Content != nil {
		content = strings.TrimSpace(*input.Content)
	}
	images := review.Images
	if input.Images != nil {
		images = *input.Images
	}

	updated, err := s.repo.UpdateByID(ctx, reviewID, repository.UpdateReviewPayload{
		Rating:  &rating,
		Title:   title,
		Content: &content,
		Images:  &images,
	})
	if err != nil {
		return ReviewResponse{}, err
	}
	if updated == nil || updated.Status == domain.ReviewStatusDeleted {
		return ReviewResponse{}, httpx.NewAppError(http.StatusNotFound, domain.ErrorCodeReviewNotFound, "Review not found", nil)
	}

	return toReviewResponse(*updated), nil
}

func (s *reviewService) DeleteReview(ctx context.Context, user domain.UserContext, reviewID string) (ReviewResponse, error) {
	if user.Role != domain.RoleCustomer {
		return ReviewResponse{}, httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "Insufficient role", nil)
	}

	review, err := s.repo.FindByID(ctx, reviewID)
	if err != nil {
		return ReviewResponse{}, err
	}
	if review == nil || review.Status == domain.ReviewStatusDeleted {
		return ReviewResponse{}, httpx.NewAppError(http.StatusNotFound, domain.ErrorCodeReviewNotFound, "Review not found", nil)
	}
	if review.BuyerID != user.UserID {
		return ReviewResponse{}, httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "You can only delete your own review", nil)
	}

	status := domain.ReviewStatusDeleted
	now := time.Now().UTC()
	updated, err := s.repo.UpdateByID(ctx, reviewID, repository.UpdateReviewPayload{
		Status:    &status,
		DeletedAt: &now,
	})
	if err != nil {
		return ReviewResponse{}, err
	}
	if updated == nil {
		return ReviewResponse{}, httpx.NewAppError(http.StatusNotFound, domain.ErrorCodeReviewNotFound, "Review not found", nil)
	}

	return toReviewResponse(*updated), nil
}

func (s *reviewService) ModerateReview(ctx context.Context, user domain.UserContext, reviewID string, input domain.ModerateReviewInput) (ReviewResponse, error) {
	if _, ok := moderatorRoles[user.Role]; !ok {
		return ReviewResponse{}, httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "Insufficient role", nil)
	}
	if input.Status == domain.ReviewStatusDeleted {
		return ReviewResponse{}, httpx.NewAppError(http.StatusBadRequest, domain.ErrorCodeBadRequest, "Moderation status cannot be DELETED", nil)
	}
	if (input.Status == domain.ReviewStatusHidden || input.Status == domain.ReviewStatusRejected) && (input.Reason == nil || strings.TrimSpace(*input.Reason) == "") {
		return ReviewResponse{}, httpx.NewAppError(http.StatusUnprocessableEntity, domain.ErrorCodeReviewModerationReasonRequire, "Moderation reason is required for HIDDEN or REJECTED status", nil)
	}

	review, err := s.repo.FindByID(ctx, reviewID)
	if err != nil {
		return ReviewResponse{}, err
	}
	if review == nil || review.Status == domain.ReviewStatusDeleted {
		return ReviewResponse{}, httpx.NewAppError(http.StatusNotFound, domain.ErrorCodeReviewNotFound, "Review not found", nil)
	}

	reason := trimStringPtr(input.Reason)
	actorID := user.UserID
	now := time.Now().UTC()
	updated, err := s.repo.UpdateByID(ctx, reviewID, repository.UpdateReviewPayload{
		Status:           &input.Status,
		ModerationReason: reason,
		ModeratedBy:      &actorID,
		ModeratedAt:      &now,
	})
	if err != nil {
		return ReviewResponse{}, err
	}
	if updated == nil || updated.Status == domain.ReviewStatusDeleted {
		return ReviewResponse{}, httpx.NewAppError(http.StatusNotFound, domain.ErrorCodeReviewNotFound, "Review not found", nil)
	}

	return toReviewResponse(*updated), nil
}

func (s *reviewService) ReplyReview(ctx context.Context, user domain.UserContext, reviewID string, input domain.ReplyReviewInput) (ReviewResponse, error) {
	if _, ok := replyRoles[user.Role]; !ok {
		return ReviewResponse{}, httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "Insufficient role", nil)
	}

	review, err := s.repo.FindByID(ctx, reviewID)
	if err != nil {
		return ReviewResponse{}, err
	}
	if review == nil || review.Status == domain.ReviewStatusDeleted {
		return ReviewResponse{}, httpx.NewAppError(http.StatusNotFound, domain.ErrorCodeReviewNotFound, "Review not found", nil)
	}
	if user.Role == domain.RoleSeller && review.SellerID != user.UserID {
		return ReviewResponse{}, httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "Seller can only reply to own product reviews", nil)
	}

	reply := &domain.ReviewReply{
		Content:   strings.TrimSpace(input.Content),
		RepliedBy: user.UserID,
		RepliedAt: time.Now().UTC(),
	}

	updated, err := s.repo.UpdateByID(ctx, reviewID, repository.UpdateReviewPayload{Reply: reply})
	if err != nil {
		return ReviewResponse{}, err
	}
	if updated == nil || updated.Status == domain.ReviewStatusDeleted {
		return ReviewResponse{}, httpx.NewAppError(http.StatusNotFound, domain.ErrorCodeReviewNotFound, "Review not found", nil)
	}

	return toReviewResponse(*updated), nil
}

func (s *reviewService) GetProductSummary(ctx context.Context, productID string) (domain.ProductSummary, error) {
	return s.repo.GetProductSummary(ctx, productID)
}

func resolveStatusFilter(user *domain.UserContext, query domain.ListReviewsQuery) (*domain.ReviewStatus, error) {
	if user == nil {
		published := domain.ReviewStatusPublished
		return &published, nil
	}

	if isModerator(user.Role) {
		return query.Status, nil
	}

	if user.Role == domain.RoleCustomer && query.BuyerID != nil && *query.BuyerID == user.UserID {
		return query.Status, nil
	}

	if query.Status != nil && *query.Status != domain.ReviewStatusPublished {
		return nil, httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "Insufficient role to query this review status", nil)
	}

	published := domain.ReviewStatusPublished
	return &published, nil
}

func assertBuyerFilterPermission(user *domain.UserContext, query domain.ListReviewsQuery) error {
	if query.BuyerID == nil {
		return nil
	}
	if user == nil {
		return httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "Authentication required for buyer filter", nil)
	}
	if isModerator(user.Role) {
		return nil
	}
	if user.Role == domain.RoleCustomer && *query.BuyerID == user.UserID {
		return nil
	}
	return httpx.NewAppError(http.StatusForbidden, domain.ErrorCodeForbidden, "Insufficient role to filter by buyerId", nil)
}

func canViewReview(user *domain.UserContext, review domain.Review) bool {
	if review.Status == domain.ReviewStatusPublished {
		return true
	}
	if review.Status == domain.ReviewStatusDeleted {
		return false
	}
	if user == nil {
		return false
	}
	if isModerator(user.Role) {
		return true
	}
	if user.Role == domain.RoleCustomer && review.BuyerID == user.UserID {
		return true
	}
	if user.Role == domain.RoleSeller && review.SellerID == user.UserID {
		return true
	}
	return false
}

func isModerator(role domain.Role) bool {
	_, ok := moderatorRoles[role]
	return ok
}

func normalizeCreateTitle(title *string) *string {
	if title == nil {
		return nil
	}
	trimmed := strings.TrimSpace(*title)
	if trimmed == "" {
		return nil
	}
	return &trimmed
}

func trimStringPtr(value *string) *string {
	if value == nil {
		return nil
	}
	trimmed := strings.TrimSpace(*value)
	return &trimmed
}

func nonNilStringSlice(items []string) []string {
	if items == nil {
		return []string{}
	}
	return items
}
